src/python/pants/option/options.py

```diff
@@ -236,15 +249,21 @@ def passthru_args_for_scope(self, scope):
     else:
       return []
 
+  def _assert_not_frozen(self):
+    if self._frozen:
+      raise self.FrozenOptionsError('cannot mutate frozen Options instance {!r}.'.format(self))
+
   def register(self, scope, *args, **kwargs):
     """Register an option in the given scope."""
+    self._assert_not_frozen()
```

Comment:
Is there anything we can do to provide a little more certainty about this freezing?

Either making an "immutable" copy of the underlying datastructures, or having a super-class which has no mutating methods and adding the mutating methods on a subclass, or something?

Reply:
definitely: I should've noted/TODO'd this in this PR iteration, but I'm planning to fast follow with a `OptionsInitializerFactory` to better cover the mutability and filesystem access aspects here. this is just a temporary first pass at that.